README.md
# lasio

[![License](http://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/kinverarity1/lasio/blob/master/LICENSE)

Read and write Log ASCII Standard files with Python.

This is a Python 2.7 and 3.3+ package to read and write Log ASCII Standard
(LAS) files, used for borehole data such as geophysical, geological, or
petrophysical logs. It's compatible with versions 1.2 and 2.0 of the LAS file
specification, published by the [Canadian Well Logging
Society](http://www.cwls.org/las). Support for LAS 3 is being worked on. In
principle it is designed to read as many types of LAS files as possible,
including ones containing common errors or non-compliant formatting.

Depending on your particular application you may also want to check out
[striplog](https://github.com/agile-geoscience/striplog) for
stratigraphic/lithological data, and
[welly](https://github.com/agile-geoscience/welly) for dealing with data at
the well level. lasio is primarily for reading & writing LAS files.

Note this is *not* a package for reading LiDAR data (also called "LAS files").

## Documentation

See here for the [complete lasio package
documentation](https://lasio.readthedocs.io/en/latest/).

## Quick start

For the minimum working requirements, you'll need numpy installed. Install
lasio with:

```bash
$ pip install lasio
```

To make sure you have everything, use this to ensure pandas, cchardet, and
openpyxl are also installed:

```bash
$ pip install lasio[all]
```

Example session:

```python
>>> import lasio
```

Data is accessible both directly as numpy arrays

```python
>>> las.keys()
['DEPT', 'DT', 'RHOB', 'NPHI', 'SFLU', 'SFLA', 'ILM', 'ILD']
>>> las['SFLU']
array([ 123.45,  123.45,  123.45, ...,  123.45,  123.45,  123.45])
>>> las['DEPT']
array([ 1670.   ,  1669.875,  1669.75 , ...,  1669.75 ,  1670.   ,
        1669.875])
```

and as ``CurveItem`` objects with associated metadata:

```python
>>> las.curves
[CurveItem(mnemonic=DEPT, unit=M, value=, descr=1  DEPTH, original_mnemonic=DEPT, data.shape=(29897,)),
CurveItem(mnemonic=DT, unit=US/M, value=, descr=2  SONIC TRANSIT TIME, original_mnemonic=DT, data.shape=(29897,)),
CurveItem(mnemonic=RHOB, unit=K/M3, value=, descr=3  BULK DENSITY, original_mnemonic=RHOB, data.shape=(29897,)),
CurveItem(mnemonic=NPHI, unit=V/V, value=, descr=4   NEUTRON POROSITY, original_mnemonic=NPHI, data.shape=(29897,)),
CurveItem(mnemonic=SFLU, unit=OHMM, value=, descr=5  RXO RESISTIVITY, original_mnemonic=SFLU, data.shape=(29897,)),
CurveItem(mnemonic=SFLA, unit=OHMM, value=, descr=6  SHALLOW RESISTIVITY, original_mnemonic=SFLA, data.shape=(29897,)),
CurveItem(mnemonic=ILM, unit=OHMM, value=, descr=7  MEDIUM RESISTIVITY, original_mnemonic=ILM, data.shape=(29897,)),
CurveItem(mnemonic=ILD, unit=OHMM, value=, descr=8  DEEP RESISTIVITY, original_mnemonic=ILD, data.shape=(29897,))]
```

Header information is parsed into simple HeaderItem objects, and stored in a
dictionary for each section of the header:

```python
>>> las.version
[HeaderItem(mnemonic=VERS, unit=, value=1.2, descr=CWLS LOG ASCII STANDARD -VERSION 1.2, original_mnemonic=VERS),
HeaderItem(mnemonic=WRAP, unit=, value=NO, descr=ONE LINE PER DEPTH STEP, original_mnemonic=WRAP)]
>>> las.well
[HeaderItem(mnemonic=STRT, unit=M, value=1670.0, descr=, original_mnemonic=STRT),
HeaderItem(mnemonic=STOP, unit=M, value=1660.0, descr=, original_mnemonic=STOP),
HeaderItem(mnemonic=STEP, unit=M, value=-0.125, descr=, original_mnemonic=STEP),
HeaderItem(mnemonic=NULL, unit=, value=-999.25, descr=, original_mnemonic=NULL),
HeaderItem(mnemonic=COMP, unit=, value=ANY OIL COMPANY LTD., descr=COMPANY, original_mnemonic=COMP),
HeaderItem(mnemonic=WELL, unit=, value=ANY ET AL OIL WELL #12, descr=WELL, original_mnemonic=WELL),
HeaderItem(mnemonic=FLD, unit=, value=EDAM, descr=FIELD, original_mnemonic=FLD),
HeaderItem(mnemonic=LOC, unit=, value=A9-16-49, descr=LOCATION, original_mnemonic=LOC),
HeaderItem(mnemonic=PROV, unit=, value=SASKATCHEWAN, descr=PROVINCE, original_mnemonic=PROV),
HeaderItem(mnemonic=SRVC, unit=, value=ANY LOGGING COMPANY LTD., descr=SERVICE COMPANY, original_mnemonic=SRVC),
HeaderItem(mnemonic=DATE, unit=, value=25-DEC-1988, descr=LOG DATE, original_mnemonic=DATE),
HeaderItem(mnemonic=UWI, unit=, value=100091604920, descr=UNIQUE WELL ID, original_mnemonic=UWI)]
>>> las.params
[HeaderItem(mnemonic=BHT, unit=DEGC, value=35.5, descr=BOTTOM HOLE TEMPERATURE, original_mnemonic=BHT),
HeaderItem(mnemonic=BS, unit=MM, value=200.0, descr=BIT SIZE, original_mnemonic=BS),
HeaderItem(mnemonic=FD, unit=K/M3, value=1000.0, descr=FLUID DENSITY, original_mnemonic=FD),
HeaderItem(mnemonic=MATR, unit=, value=0.0, descr=NEUTRON MATRIX(0=LIME,1=SAND,2=DOLO), original_mnemonic=MATR),
HeaderItem(mnemonic=MDEN, unit=, value=2710.0, descr=LOGGING MATRIX DENSITY, original_mnemonic=MDEN),
HeaderItem(mnemonic=RMF, unit=OHMM, value=0.216, descr=MUD FILTRATE RESISTIVITY, original_mnemonic=RMF),
HeaderItem(mnemonic=DFD, unit=K/M3, value=1525.0, descr=DRILL FLUID DENSITY, original_mnemonic=DFD)]
```

The data is stored as a 2D numpy array:

```python
>>> las.data
array([[ 1670.   ,   123.45 ,  2550.   , ...,   123.45 ,   110.2  ,   105.6  ],
       [ 1669.875,   123.45 ,  2550.   , ...,   123.45 ,   110.2  ,   105.6  ],
       [ 1669.75 ,   123.45 ,  2550.   , ...,   123.45 ,   110.2  ,   105.6  ],
       ...,
       [ 1669.75 ,   123.45 ,  2550.   , ...,   123.45 ,   110.2  ,   105.6  ],
       [ 1670.   ,   123.45 ,  2550.   , ...,   123.45 ,   110.2  ,   105.6  ],
       [ 1669.875,   123.45 ,  2550.   , ...,   123.45 ,   110.2  ,   105.6  ]])
```

You can also retrieve and load data as a ``pandas`` DataFrame, build LAS files
from scratch, write them back to disc, and export to Excel, amongst other
things.

See the [package documentation](https://lasio.readthedocs.io/en/latest/) for
more details.

## License

MIT


CHANGELOG.md
# List of changes

## Version 0.25
- Add stack_curves() method to allow joining a set of curves into a 2D array (issue #284, PR #293)
- Add lasio.examples module (#296)
- Fix #278 (leading zeroes were being stripped from API/UWI numbers)
- Fix #286 (error on trying to write a file with one row of data)
- Fix #258 (do not catch Ctrl+C when reading file)
- Fix #292 (improve error checking for when trying to write non-2D data)
- Fix #277 (allow pathlib objects to lasio.read)
- Fix #264 (allow periods in mnemonics to be retained in specific cases)
- Fix #201 (adjust descr parsing in \~P section to allow times in the descr, see PR #298)
- Fix #302 (change in str(datetime) handling)
- Fixes to JSON output (#300, #303)
- Fix #304 (add column_fmt argument to LASFile.write method)

## Version 0.24

- Fix #256 (parse units in brackets and add index_unit kwarg)

## Version 0.23

- Fix #259 (error when encoding missing from URL response headers)
- Fix #262 (broken build due to cchardet dependency)

## Version 0.22

- Fix #252 (removing case sensitivity in index_unit checks)
- Fix #249 (fix bug producing df without converting to floats)
- Attempt to fix Lasso classification on GitHub

## Version 0.21

- Fix #236 and #237 (can now read ASCII in ~Data section)
- Fix #239 (Petrel can't read lasio output)

## Version 0.20

- Fix #233 (pickling error lost Curve.data during multiprocessing)
- Fix #226 (do not issue warning on empty ~Parameter section)
- Revised default behaviour to using null_policy='strict' (ref. #227)
- Fix #221 (depths > 10000 were being rounded by default)
- Fix #225 (file handle leaked if exception during parsing)

## Version 0.19

- Fix #223 (critical version/installation bug)

## Version 0.18

- Fix version numbering setup
- Fix #92 (can ignore blah blah lines in ~C section)
- Fix #209 (can now add curves with LASFile['mnemonic'] = [1, 2, 3])
- Fix #213 (LASFile.data is now a lazily generated property, with setter)
- Fix #218 (LASFile.append_curve was not adding data=[...] properly)
- Fix #216 (LASFile now raises KeyError for missing mnemonics)
- Fix #214 (first duplicate mnemonic when added was missing the :1)

## Version 0.17

- Add Appveyor continuous integration testing
- Add example notebook for how to use python logging module
- Fix #160 (add methods to LASFile for inserting curves)
- Fix #155 (implement del keyword for header items)
- Fix #142 (implement slicing for SectionItems)
- Fix #135 (UWI numbers losing their leading zeros)
- Fix #153 (fix SectionItems pprint repr in Python 3)
- Fix #81 (accept header items with missing colon)
- Fix #71 (add Docker build for lasio to DockerHub)
- Fix #210 (allow upper/lowercase standardization of mnemonics on read)
- Document recent additions (nearly up to date) (in Sphinx docs)

## Version 0.16

- Add read_policy and null_policy keywords - see documentation for details
- Fix bugs around files with missing ~V ~W ~P or ~C sections (#84 #85 #78)
- Fix #17 involving files with commas as a decimal mark
- Improve LASHeaderError traceback message
- Fix bug involving files with ~A but no data lines following
- Fix bug with blank line at start of file
- Fix bug involving missing or duplicate STRT, STOP and STEP mnemonics

## Version 0.15.1

- Major performance improvements with both memory and speed
- Major improvement to read parser, now using iteration
- Add ``LASFile.to_excel()`` and ``LASFile.to_csv()`` export methods
- Improve ``las2excelbulk.py`` script
- Published new and updated Sphinx documentation
- Improved character encoding handling when ``chardet`` not installed
- ``autodetect_encoding=True`` by default
- Allow reading of multiple non-standard header sections (#167, #168)
- Add flexibility in reading corrupted headers (``ignore_header_errors=True``)
- Add ability to avoid reading in data (``ignore_data=True``)
- Remove excessive debugging messages
- Fix bug #164 where ``FEET`` was not recognised as ``FT``
- Fix major globals() bug #141 affecting LASFile.add_curve
- Add command-line version script ``$ lasio`` to show version number.

Version 0.14 and 0.15 skipped due to broken PyPI upload.

## Version 0.13

- Other minor bug fixes inc inability to rename mnemonics in written LAS file.

## Version 0.11.2

- Fix bug with not correctly figuring out units for LASFile.write()
- Add ``LASFile.add_curve(CurveItem)`` method which automatically goes to the old
  method at ``LASFile.add_curve_raw(mnemonic=, data=, ...)`` if necessary, so it
  should be transparent to users

## Version 0.11

- Reorganise code into modules
- various

## Version 0.10

- Internal change to SectionItems for future LAS 3.0 support
- Added JSON encoder
- Added examples for using pandas DataFrame (.df attribute)
- LAS > Excel script refined (las2excel.py)

## Version 0.9.1 (2015-11-11)

 - pandas.DataFrame now as .df attribute, bugfix

## Version 0.8 (2015-08-20)

 - numerous bug fixes, API documentation added

## Version 0.7 (2015-08-08)

 - all tests passing on Python 2.6 through 3.4

## Version 0.6 (2015-08-05)

 - bugfixes and renamed from ``las_reader`` to ``lasio``

## Version 0.5 (2015-08-01)

 - Improvements to writing LAS files

## Version 0.4 (2015-07-26)

 - Improved handling of character encodings, other internal improvements

## Version 0.3 (2015-07-23)

 - Added Python 3 support, now reads LAS 1.2 and 2.0

## Version 0.2 (2015-07-08)

 - Tidied code and published on PyPI

lasio/defaults.py
import re

import numpy as np

from .las_items import (
    HeaderItem, SectionItems, OrderedDict
)


def get_default_items():
    return {
        'Version': SectionItems([
            HeaderItem('VERS', '', 2.0, 'CWLS log ASCII Standard -VERSION 2.0'),
            HeaderItem('WRAP', '', 'NO', 'One line per depth step'),
            HeaderItem('DLM', '', 'SPACE', 'Column Data Section Delimiter'),
        ]),
        'Well': SectionItems([
            HeaderItem('STRT', 'm', np.nan, 'START DEPTH'),
            HeaderItem('STOP', 'm', np.nan, 'STOP DEPTH'),
            HeaderItem('STEP', 'm', np.nan, 'STEP'),
            HeaderItem('NULL', '', -9999.25, 'NULL VALUE'),
            HeaderItem('COMP', '', '', 'COMPANY'),
            HeaderItem('WELL', '', '', 'WELL'),
            HeaderItem('FLD', '', '', 'FIELD'),
            HeaderItem('LOC', '', '', 'LOCATION'),
            HeaderItem('PROV', '', '', 'PROVINCE'),
            HeaderItem('CNTY', '', '', 'COUNTY'),
            HeaderItem('STAT', '', '', 'STATE'),
            HeaderItem('CTRY', '', '', 'COUNTRY'),
            HeaderItem('SRVC', '', '', 'SERVICE COMPANY'),
            HeaderItem('DATE', '', '', 'DATE'),
            HeaderItem('UWI', '', '', 'UNIQUE WELL ID'),
            HeaderItem('API', '', '', 'API NUMBER')
        ]),
        'Curves': SectionItems([]),
        'Parameter': SectionItems([]),
        'Other': '',
        'Data': np.zeros(shape=(0, 1)),
    }


ORDER_DEFINITIONS = {
    1.2: OrderedDict([
        ('Version', ['value:descr']),
        ('Well', [
            'descr:value',
            ('value:descr', ['STRT', 'STOP', 'STEP', 'NULL'])]),
        ('Curves', ['value:descr']),
        ('Parameter', ['value:descr']),
    ]),
    2.0: OrderedDict([
        ('Version', ['value:descr']),
        ('Well', ['value:descr']),
        ('Curves', ['value:descr']),
        ('Parameter', ['value:descr'])
    ])}

DEPTH_UNITS = {
    'FT': ("FT", "F", "FEET", "FOOT"),
    'M': ("M", "METER", "METERS", "METRE", "METRES"),
    }

READ_POLICIES = {
    'default': ['comma-decimal-mark', 'run-on(-)', 'run-on(.)', 'run-on(NaN.)'],
    }

READ_SUBS = {
    'comma-decimal-mark': [(re.compile(r'(\d),(\d)'), r'\1.\2'), ],
    'run-on(-)': [(re.compile(r'(\d)-(\d)'), r'\1 -\2'), ],
    'run-on(.)': [(re.compile(r'-?\d*\.\d*\.\d*'), ' NaN NaN '), ],
    'run-on(NaN.)': [(re.compile(r'NaN[\.-]\d+'), ' NaN NaN '), ],
    }

NULL_POLICIES = {
    'none': [],
    'strict': ['NULL', ],
    'common': ['NULL', '(null)', '-',
               '9999.25', '999.25', 'NA', 'INF', 'IO', 'IND'],
    'aggressive': ['NULL', '(null)', '--',
                   '9999.25', '999.25', 'NA', 'INF', 'IO', 'IND',
                   '999', '999.99', '9999', '9999.99' '2147483647', '32767',
                   '-0.0', ],
    'all': ['NULL', '(null)', '-',
            '9999.25', '999.25', 'NA', 'INF', 'IO', 'IND',
            '999', '999.99', '9999', '9999.99' '2147483647', '32767', '-0.0',
            'numbers-only', ],
    'numbers-only': ['numbers-only', ]
    }

NULL_SUBS = {
    'NULL': [None, ],       # special case to be handled in LASFile.read()
    '999.25': [-999.25, 999.25],
    '9999.25': [-9999.25, 9999.25],
    '999.99': [-999.99, 999.99],
    '9999.99': [-9999.99, 9999.99],
    '999': [-999, 999],
    '9999': [-9999, 9999],
    '2147483647': [-2147483647, 2147483647],
    '32767': [-32767, 32767],
    '(null)': [(re.compile(r' \(null\)'), ' NaN'),
               (re.compile(r'\(null\) '), 'NaN '),
               (re.compile(r' \(NULL\)'), ' NaN'),
               (re.compile(r'\(NULL\) '), 'NaN '),
               (re.compile(r' null'), ' NaN'),
               (re.compile(r'null '), 'NaN '),
               (re.compile(r' NULL'), ' NaN'),
               (re.compile(r'NULL '), 'NaN '), ],
    '-': [(re.compile(r' -+ '), ' NaN '), ],
    'NA': [(re.compile(r'(#N/A)[ ]'), 'NaN '),
           (re.compile(r'[ ](#N/A)'), ' NaN'), ],
    'INF': [(re.compile(r'(-?1\.#INF)[ ]'), 'NaN '),
            (re.compile(r'[ ](-?1\.#INF[0-9]*)'), ' NaN'), ],
    'IO': [(re.compile(r'(-?1\.#IO)[ ]'), 'NaN '),
           (re.compile(r'[ ](-?1\.#IO)'), ' NaN'), ],
    'IND': [(re.compile(r'(-?1\.#IND)[ ]'), 'NaN '),
            (re.compile(r'[ ](-?1\.#IND[0-9]*)'), ' NaN'), ],
    '-0.0': [(re.compile(r'(-0\.0)[ ]'), 'NaN '),
             (re.compile(r'[ ](-0\.0)'), ' NaN'), ],
    'numbers-only': [(re.compile(r'([^ 0-9.\-+]+)[ ]'), 'NaN '),
                     (re.compile(r'[ ]([^ 0-9.\-+]+)'), ' NaN'), ],
    }

